File: lib/iriq/cross_host_shape.rb
require "set"

module Iriq
  # A route shape that recurs across multiple hosts.
  #
  # Emitted by Corpus#cross_host_shapes. The shape string ("/users/{user_id}")
  # is the cluster's rendered placeholder form; two clusters with the same
  # shape but different hosts coalesce into one CrossHostShape record.
  #
  # A shape appearing at N hosts is strong evidence of a semantic pattern
  # rather than a host-local quirk — independent hosts are unlikely to
  # invent the same `/users/{integer}` structure by accident. Future work
  # can feed this signal into proposal confidence and corpus-informed
  # normalization (raise weight when a Shape has cross-host support).
  class CrossHostShape
    attr_reader :shape, :hosts, :observation_count

    def initialize(shape:, hosts:, observation_count:)
      @shape             = shape
      @hosts             = hosts.is_a?(Set) ? hosts.dup.freeze : Set.new(hosts).freeze
      @observation_count = observation_count
    end

    def host_count
      @hosts.size
    end

    def to_h
      {
        shape:             @shape,
        hosts:             @hosts.to_a.sort,
        host_count:        host_count,
        observation_count: @observation_count,
      }
    end
  end
end

File: rust/iriq/src/cross_host_shape.rs
use crate::corpus::Corpus;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct CrossHostShape {
    pub shape: String,
    pub hosts: Vec<String>,
    pub observation_count: usize,
}

impl CrossHostShape {
    pub fn host_count(&self) -> usize {
        self.hosts.len()
    }
}

/// Lists route shapes (the path-only form, stripped of host) that recur
/// across multiple hosts. Mirrors Go's Corpus.CrossHostShapes.
pub fn cross_host_shapes(cp: &Corpus, min_hosts: usize) -> Vec<CrossHostShape> {
    let min = if min_hosts == 0 { 2 } else { min_hosts };
    let mut by_shape: HashMap<String, (Vec<String>, usize)> = HashMap::new();
    for c in cp.clusters() {
        let entry = by_shape
            .entry(c.shape.clone())
            .or_insert_with(|| (Vec::new(), 0));
        if !entry.0.contains(&c.host) {
            entry.0.push(c.host.clone());
        }
        entry.1 += c.count;
    }
    let mut out: Vec<CrossHostShape> = by_shape
        .into_iter()
        .filter_map(|(shape, (mut hosts, count))| {
            if hosts.len() < min {
                None
            } else {
                hosts.sort();
                Some(CrossHostShape {
                    shape,
                    hosts,
                    observation_count: count,
                })
            }
        })
        .collect();
    out.sort_by(|a, b| {
        b.host_count()
            .cmp(&a.host_count())
            .then(b.observation_count.cmp(&a.observation_count))
            .then(a.shape.cmp(&b.shape))
    });
    out
}
